libnetwork/sandbox_dns_unix.go

```diff
@@ -28,10 +28,24 @@ const (
 	resolverIPSandbox = "127.0.0.11"
 )
 
-// finishInitDNS is to be called after the container namespace has been created,
-// before it the user process is started. The container's support for IPv6 can be
-// determined at this point.
-func (sb *Sandbox) finishInitDNS(ctx context.Context) error {
+// AddHostsEntry adds an entry to /etc/hosts.
+func (sb *Sandbox) AddHostsEntry(ctx context.Context, name, ip string) error {
+	sb.config.extraHosts = append(sb.config.extraHosts, extraHost{name: name, IP: ip})
+	return sb.rebuildHostsFile(ctx)
```

Comment:
`Sandbox.AddHostsEntry` is only used for legacy `--link` options for containers on the default bridge network, before they're started - and those containers are only likely to be connected to the default bridge (only one Endpoint) ... so I don't think there's much to gain by optimising it.

`Sandbox.rebuildHostsFile` is called from `SetKey` for a newly created container once the its support for IPv6 is known. That container could be connected to multiple endpoints - so maybe there are some msecs to be gained by accumulating addresses from all of the endpoints. But, optimising the whole `etchosts` thing so that it doesn't keep writing/reading/merging the contents (before a container starts and the user's had a chance to modify it), is probably the way to go if that's an issue.

These `/etc/hosts` entries are pretty questionable anyway for user-defined networks that have an internal DNS resolver. If we figure out a [fix for buildkit](https://github.com/moby/moby/pull/48020), so that we can use the internal resolver for the default network too, even more so.

I don't really want to muck about with it in this PR, maybe as a follow-up if needed?

Reply:
This is now https://github.com/moby/moby/pull/48823